### PR TITLE
EZP-31290: Removed the required '@' prefix on custom view matchers

### PR DESCRIPTION
**Pull request by bdunogier**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31290](https://jira.ez.no/browse/EZP-31290)
| **Bug/Improvement**| bug
| **New feature**    | yes/no
| **Target version** | `8.x` / `master`
| **BC breaks**      | yes, but no
| **Tests pass**     | maybe
| **Doc needed**     | maybe

Custom, service based view matchers were changed in kernel v8 to require an `@` prefix. It turns out that it wasn't necessary in earlier versions, as the matcher's name would be [searched for as is in the container](https://github.com/ezsystems/ezpublish-kernel/blob/7.5/eZ/Bundle/EzPublishCoreBundle/Matcher/ServiceAwareMatcherFactory.php#L28-L30). This change removes the check & handling of that prefix.

### CODE REVIEW

File: eZ/Bundle/EzPublishCoreBundle/Matcher/ServiceAwareMatcherFactory.php [hunk 1]
```diff
@@ -38,10 +38,10 @@ public function __construct(
      */
     protected function getMatcher($matcherIdentifier)
     {
-        if (strpos($matcherIdentifier, '@') === 0) {
-            return $this->viewMatcherRegistry->getMatcher(substr($matcherIdentifier, 1));
+        try {
+            return $this->viewMatcherRegistry->getMatcher($matcherIdentifier);
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Matcher/ServiceAwareMatcherFactory.php]: Matchers should actually be rather services and fully utilize Dependency Injection. If you want short identifiers, those should be literals associated with a given matcher (obtained either via method fulfilling a contract or as an attribute of tagged service (like `alias` in many other cases).

There's an underlying issue in [that portion of code](https://github.com/ezsystems/ezpublish-kernel/blob/v8.0.0-beta3/eZ/Publish/Core/MVC/Symfony/SiteAccess/MatcherBuilder.php#L29-L42) and [that](https://github.com/ezsystems/ezpublish-kernel/blob/dc7e85bb489e5e9b0c188701933da4b4929dca8f/eZ/Publish/Core/MVC/Symfony/Matcher/ClassNameMatcherFactory.php#L73-L95). Those should also utilize DI to get services.

In other words, the change could go in, provided we:
1. Introduce `identifier` for matchers.
2. Get rid of the exception-driven logic here (should always throw).
3. Drop dynamic namespace-based instantiation of matchers and register all of them as services.
4. Move a responsibility of getting matchers entirely to the registry (3. should solve that).

**Review by alongosz — CHANGES_REQUESTED** [review]
Hi. I would go for this, but other things need to be changed as well. It's also worth to mention that it's inconsistent with the convention used for SiteAccess matches, so that one would have to be changed as well.
What I would need here:

**Comment by alongosz** [comment]
POV ping @adamwojs @mikadamczyk @Nattfarinn @webhdx @ViniTou  about what I've said [above](https://github.com/ezsystems/ezpublish-kernel/pull/2916#pullrequestreview-343089659).

File: eZ/Bundle/EzPublishCoreBundle/Matcher/ServiceAwareMatcherFactory.php (hunk 1 above)
Comment (bdunogier) [reply eZ/Bundle/EzPublishCoreBundle/Matcher/ServiceAwareMatcherFactory.php]: Nothing to oppose to that.

The most important aspects for me, by order of priority:
- provide a clear, consistent syntax for _using_ matchers in config, independently of them being core or 3rd party, classes or services
- respect BC for those, as there are implementations around (mostly Netgen's, including Tags)

That aside, it doesn't bother me to make them all services. I can't recall the reasons why we chose to implement them that way. Maybe @lolautruche could shed some light ?

**Comment by bdunogier** [comment]
> Hi. I would go for this, but other things need to be changed as well.
Do they _need_ to be changed, or _should_ they be changed ?

> It's also worth to mention that it's inconsistent with the convention used for SiteAccess matches, so that one would have to be changed as well.
I think that custom siteaccess matchers are far less common than view matchers. I would indeed unify them to behave like view matchers (without the `@`). The most important criteria here is that it is consistent with 2.x when it comes to view matchers, I believe. That PR alone does make it consistent, does it not ?

**Comment by alongosz** [comment]
>> Hi. I would go for this, but other things need to be changed as well.

> Do they need to be changed, or should they be changed ?

Given the need for exception-driven logic I'd say they need to be changed, the solution seems incomplete for me right now.

>> It's also worth to mention that it's inconsistent with the convention used for SiteAccess matches, so that one would have to be changed as well.

> I think that custom siteaccess matchers are far less common than view matchers. I would indeed unify them to behave like view matchers (without the @). The most important criteria here is that it is consistent with 2.x when it comes to view matchers, I believe. That PR alone does make it consistent, does it not ?

Yeah, actually SA matchers should be a separate effort anyway. Still, would be good to have it unified within one release.

**Comment by bdunogier** [comment]
Fine, I won't fight over this :-)

My only _current_ concern is that in the current state, the query field will not work as documented. Maybe I'll just go for one of the workarounds I mentioned earlier in the query field.
